runtime/src/main/cpp/Memory.cpp

```diff
@@ -627,51 +627,66 @@ void ScanRoots(MemoryState*);
 void CollectRoots(MemoryState*);
 
 template<bool useColor>
-void MarkGray(ContainerHeader* container) {
-  if (useColor) {
-    if (container->color() == CONTAINER_TAG_GC_GRAY) return;
-  } else {
-    if (container->marked()) return;
-  }
-  if (useColor) {
-    container->setColor(CONTAINER_TAG_GC_GRAY);
-  } else {
-    container->mark();
-  }
-  traverseContainerReferredObjects(container, [](ObjHeader* ref) {
-    auto childContainer = ref->container();
-    RuntimeAssert(!isArena(childContainer), "A reference to local object is encountered");
-
-    if (!childContainer->permanentOrFrozen()) {
-      childContainer->decRefCount<false>();
-      MarkGray<useColor>(childContainer);
+void MarkGray(ContainerHeader* start) {
+  ContainerHeaderDeque toVisit;
+  toVisit.push_back(start);
+
+  while (!toVisit.empty()) {
+    auto container = toVisit.front();
+    toVisit.pop_front();
+    if (useColor) {
+      if (container->color() == CONTAINER_TAG_GC_GRAY) continue;
+    } else {
+       if (container->marked()) continue;
     }
-  });
+    if (useColor) {
+      container->setColor(CONTAINER_TAG_GC_GRAY);
+    } else {
+      container->mark();
+    }
+    traverseContainerReferredObjects(container, [&toVisit](ObjHeader* ref) {
+      auto childContainer = ref->container();
+      RuntimeAssert(!isArena(childContainer), "A reference to local object is encountered");
+
+      if (!childContainer->permanentOrFrozen()) {
+        childContainer->decRefCount<false>();
+        toVisit.push_front(childContainer);
```

Comment:
Is it important we do DFS, not BFS here?

Reply:
Maybe check that `childContainer` is marked before adding to the queue?

Reply:
Maybe check that `childContainer` is marked before adding it to the queue?